### ISSUE TITLE
Allow users to take a photo while filling in an xform in Enketo and upload the photo with the form

### ISSUE THREAD

**Issue by sglangevin** [medic/medic-webapp#3096]
Add support to our implementation of Enketo for taking a photo within a form and uploading the photo with the form. This is important for tracking use of rapid tests for malaria, ensuring that they are actually being used and that the results are being read correctly by CHWs. It's also applicable to projects that track usage of particular commodities, such as a project tracking the use of malaria treatments.

**Comment by alxndrsn** [comment]
I'm reviewing the backlog of enketo-specific issues.  Does this feature have requirement in a current or planned project?

**Comment by sglangevin** [comment]
Yes - LG has been asking for this for several months. #1554 is higher priority than this issue is, though.

**Comment by alxndrsn** [comment]
N.B. initial proposed use is:
> take photos of malaria rapid tests to help with auditing

**Comment by alxndrsn** [comment]
Example XForm:

```
<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:ev="http://www.w3.org/2001/xml-events" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:jr="http://openrosa.org/javarosa">
  <h:head>
    <h:title>Image Widget</h:title>
    
    <model>
      <instance>
      	<widgets xmlns="widgets">
    	  <my_image/>
  		</widgets>
      </instance>
    </model>

    <bind nodeset="/widgets/my_image" type="binary"/>
  </h:head>

  <h:body>   
    <upload ref="my_image" mediatype="image/*">
      <hint>Take or select a picture</hint>
      <label>Image widget</label>
    </upload>
  </h:body>
</h:html>
```
---

The following HTML works out of the box in the standard android browser.

```
<input type="file" accept="image/*;capture=camera">
```

However, it will need custom handling in both `WebView` and `XWalkView`, e.g. 
* https://stackoverflow.com/questions/5803914/can-i-open-camera-in-webview
* https://stackoverflow.com/questions/15805886/android-webview-upload-file-and-camera
* https://stackoverflow.com/questions/18568566/webview-input-of-type-file-camera-and-image

**Comment by sglangevin** [comment]
@alxndrsn following up on this one - it's marked as active work, just curious if we should keep it there or delay it a bit.

**Comment by alxndrsn** [comment]
If it's not a priority, I'll leave it on the back burner.  It's the kind of work I'll come back to if e.g. I get stuck without internet, or have a spare afternoon.

**Comment by sglangevin** [comment]
It's not required for v2.12, but I'd like to come back to it for v2.13, if you don't get to it sooner.

**Comment by alxndrsn** [comment]
Ref https://github.com/opendatakit/xforms-spec/issues/112

**Comment by garethbowen** [comment]
@sglangevin Should this be scheduled in a Project? Maybe 2.13.0?

**Comment by sglangevin** [comment]
@alxndrsn is this done? If so, we can include it in 2.13.0. If not, it can wait longer as 2.13.0 has a lot in it.

**Comment by alxndrsn** [comment]
It should probably wait - i've got it working, but not production-ready.  The following are still to do:
- [ ] finalise photo/camera/media-related changes to xforms spec and update the following:
  - [ ] our XSL stylesheets
  - [ ] get medic-android intent handler
  - [ ] update enketo photo widget
- [ ] write some tests

**Comment by garethbowen** [comment]
@alxndrsn Back to you. Once those tests are green, merge away!

**Comment by ngaruko** [comment]
Any clue on how to AT this @sglangevin ?

**Comment by alxndrsn** [comment]
@ngaruko there's a demo Xform in the `demo-forms` directory in `medic-webapp` which should help.

**Comment by sglangevin** [comment]
Assigning to myself for AT.

**Comment by sglangevin** [comment]
@alxndrsn I've uploaded the demo form to alpha.dev and tried it on v0.3.9 of the Android app. I see the widget that says "select file" but when I click on it, nothing happens. I'm not seeing any errors in the console either. Any idea what could be going on?

**Comment by alxndrsn** [comment]
Checking this...

**Comment by alxndrsn** [comment]
@sglangevin works for me on Samsung J1 Mini; which handset are you having problems with?

**Comment by sglangevin** [comment]
I tried on alpha.dev on Tecno Y4 and on an LG LS660, both Android 4.4.2 and Android app version 0.3.9. In both cases, no console error but clicking "Choose File" does nothing.

**Comment by garethbowen** [comment]
@sglangevin Have you looked in the android logs? It might be an issue on the app side.

**Comment by alxndrsn** [comment]
I'll test on 4.4.2 today.

**Comment by sglangevin** [comment]
I tested a snapshot apk on my Tecno Y4 and it was working well. This will need additional testing once the code is merged and a new version of medic-android is released.

**Comment by alxndrsn** [comment]
* [ ] add maximum size attribute to file inputs in enketo forms (from xforms?  variable?  etc.)
* [ ] downsize big images in java before sending to the browser
* [ ] revert to upstream ImagePicker

**Comment by sglangevin** [comment]
@alxndrsn what's the status on this one? is it far enough along that i can re-test on y4?

**Comment by garethbowen** [comment]
Back to you Alex!

**Comment by garethbowen** [comment]
LG! Feel free to merge and backport.

**Comment by sglangevin** [comment]
Tested on both 2.14.0-beta.4 and on alpha on my Tecno tablet. App version 0.4.0, Android version 5.1 and I don't get the image widget:
![screenshot_2017-11-20-17-49-07](https://user-images.githubusercontent.com/1286381/33045572-30f690ac-ce02-11e7-8537-1d24eb663dc3.png)

any ideas @alxndrsn?

**Comment by sglangevin** [comment]
It's also no longer working on my Tecno, Android app v0.1.157, webapp v2.14.0-beta.4.

**Comment by alxndrsn** [comment]
Please see https://github.com/medic/medic-android/releases - you'll need at least `v0.3.12`.

**Comment by sglangevin** [comment]
Per my comment: https://github.com/medic/medic-webapp/issues/3096#issuecomment-345860405 I was testing using 0.4.0.

**Comment by alxndrsn** [comment]
Working fine for me on Tecno Y4 webapp v`2.14-beta.4` with android app v`0.4.0` and the photo form at https://github.com/medic/medic-webapp/blob/master/demo-forms/photos-on-android.xml.

**Comment by alxndrsn** [comment]
Looks like this is proguard's fault.
1. [ ] test locally with proguard enabled
2. [ ] add Make target so that testing with proguard on all future code is simple (and therefore happens)

**Comment by alxndrsn** [comment]
I've tested this locally with a proguard-enabled APK on Y4 and Samsung J1 Mini Prime, and both work fine.  This will need further testing with APK from app store, and checking if there is anything obvious in the app logs.

**Comment by sglangevin** [comment]
Working on medic-android v0.4.3 and 2.14.0-beta.6. However, the app seems to crash easily when using this feature. Not sure if it's just the tablet I'm testing on or not. @alxndrsn I'll send you the logs I collected today and I'll also try testing on another device on Monday.

**Comment by sglangevin** [comment]
Working well on medic-android v0.4.4 and 2.14.0-beta.6. Moving to ready!
